pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -336,6 +364,76 @@ func (m *ManagerImpl) Stop() error {
 	return nil
 }
 
+func (m *ManagerImpl) addEndpointProbeMode(resourceName string, socketPath string) (error, chan bool) {
+	chanForAckOfNotification := make(chan bool)
+	existingDevs := make(map[string]pluginapi.Device)
+	m.mutex.Lock()
+	old, ok := m.endpoints[resourceName]
+	if ok && old != nil {
+		// Pass devices of previous endpoint into re-registered one,
+		// to avoid potential orphaned devices upon re-registration
+		devices := make(map[string]pluginapi.Device)
+		for _, device := range old.getDevices() {
+			device.Health = pluginapi.Unhealthy
+			devices[device.ID] = device
+		}
+		existingDevs = devices
+	}
+	m.mutex.Unlock()
+
+	e, err := newEndpointImpl(socketPath, resourceName, existingDevs, m.callback)
+	if err != nil {
+		glog.Errorf("Failed to dial device plugin with socketPath %s: %v", socketPath, err)
+		return fmt.Errorf("Failed to dial device plugin with socketPath %s: %v", socketPath, err), chanForAckOfNotification
+	}
+
+	m.mutex.Lock()
+	options, err := e.client.GetDevicePluginOptions(context.Background(), &pluginapi.Empty{})
+	if err != nil {
+		glog.Errorf("Failed to get device plugin options: %v", err)
+		return fmt.Errorf("Failed to get device plugin options: %v", err), chanForAckOfNotification
+	}
+	m.pluginOpts[resourceName] = options
+	// Check for potential re-registration during the initialization of new endpoint,
+	// and skip updating if re-registration happens.
+	// TODO: simplify the part once we have a better way to handle registered devices
+	ext := m.endpoints[resourceName]
+	if ext != old {
+		glog.Warningf("Some other endpoint %v is added while endpoint %v is initialized", ext, e)
+		m.mutex.Unlock()
+		e.stop()
+		return fmt.Errorf("Some other endpoint %v is added while endpoint %v is initialized", ext, e), chanForAckOfNotification
+	}
+	// Associates the newly created endpoint with the corresponding resource name.
+	// Stops existing endpoint if there is any.
+	m.endpoints[resourceName] = e
+	glog.V(2).Infof("Registered endpoint %v", e)
+	m.mutex.Unlock()
+
+	if old != nil {
+		old.stop()
+	}
```

Comment:
A big part of this function looks similar to addEndpoint(). Is there any way we may take the common part into a common func?